lib/std/compress/deflate/compressor.zig

```diff
@@ -0,0 +1,1133 @@
+        },
+        Compression.level_8 => return .{
+            .good = 32,
+            .lazy = 128,
+            .nice = 258,
+            .chain = 1024,
+            .fast_skip_hashshing = skip_never,
+        },
+        Compression.best_compression => return .{
+            .good = 32,
+            .lazy = 258,
+            .nice = 258,
+            .chain = 4096,
+            .fast_skip_hashshing = skip_never,
+        },
+    }
+}
+
+// matchLen returns the number of matching bytes in a and b
+// up to length 'max'. Both slices must be at least 'max'
+// bytes in size.
+fn matchLen(a: []u8, b: []u8, max: u32) u32 {
+    var bounded_a = a[0..max];
+    var bounded_b = b[0..max];
+    for (bounded_a) |av, i| {
+        if (bounded_b[i] != av) {
+            return @intCast(u32, i);
+        }
+    }
+    return max;
+}
+
+const hash_mul = 0x1e35a7bd;
+
+// hash4 returns a hash representation of the first 4 bytes
+// of the supplied slice.
+// The caller must ensure that b.len >= 4.
+fn hash4(b: []u8) u32 {
+    return ((@intCast(u32, b[3]) |
+        @intCast(u32, b[2]) << 8 |
+        @intCast(u32, b[1]) << 16 |
+        @intCast(u32, b[0]) << 24) *% hash_mul) >> (32 - hash_bits);
+}
+
+// bulkHash4 will compute hashes using the same
+// algorithm as hash4
+fn bulkHash4(b: []u8, dst: []u32) u32 {
+    if (b.len < min_match_length) {
+        return 0;
+    }
+    var hb =
+        @intCast(u32, b[3]) |
+        @intCast(u32, b[2]) << 8 |
+        @intCast(u32, b[1]) << 16 |
+        @intCast(u32, b[0]) << 24;
+
+    dst[0] = (hb *% hash_mul) >> (32 - hash_bits);
+    var end = b.len - min_match_length + 1;
+    var i: u32 = 1;
+    while (i < end) : (i += 1) {
+        hb = (hb << 8) | @intCast(u32, b[i + 3]);
+        dst[i] = (hb *% hash_mul) >> (32 - hash_bits);
+    }
+
+    return hb;
+}
+
+// Returns a new Compressor compressing data at the given level.
+// Following zlib, levels range from 1 (best_speed) to 9 (best_compression);
+// higher levels typically run slower but compress more. Level 0
+// (no_compression) does not attempt any compression; it only adds the
+// necessary DEFLATE framing.
+// Level -1 (default_compression) uses the default compression level.
+// Level -2 (huffman_only) will use Huffman compression only, giving
+// a very fast compression for all types of input, but sacrificing considerable
+// compression efficiency.
+//
+// If level is in the range [-2, 9] then the error returned will be nil.
+// Otherwise the error returned will be non-nil.
+//
+// `dictionary` is optional and initializes the new
+// `Compressor` with a preset dictionary. The returned Compressor behaves
+// as if the dictionary had been written to it without producing
+// any compressed output. The compressed data written to hm_bw
+// can only be decompressed by a Decompressor initialized with the
+// same dictionary.
+pub fn compressor(
+    allocator: Allocator,
+    writer: anytype,
+    level: Compression,
+    dictionary: ?[]const u8,
```

Comment:
This arguments should be put in a `CompressorOptions` struct with defaults of `default_compression` and `null`.

Reply:
`CompressorOptions` added